File: mobile/src/services/feedback.ts
import { api } from './api';
import { logger } from '../utils/logger';

export interface FeedbackRequest {
  feedback_type: 'positive' | 'negative' | 'correction';
  feedback_context: 'response' | 'suggestion' | 'memory_retrieval' | 'action';
  conversation_id?: string;
  message_id?: string;
  user_query?: string;
  ai_response?: string;
  correction_text?: string;
  memories_used?: string[];
}

export interface FeedbackResponse {
  id: string;
  feedback_type: string;
  message: string;
}

export interface UserPreferences {
  preferences: Record<string, any>;
  user_model_prompt: string;
}

export interface LearningStats {
  memories: {
    total: number;
    average_strength: number;
    average_emotional_weight: number;
  };
  feedback: {
    positive: number;
    negative: number;
    corrections: number;
  };
  preferences_learned: number;
  insights_generated: number;
}

class FeedbackService {
  /**
   * Submit feedback on an AI response
   */
  async submitFeedback(request: FeedbackRequest): Promise<FeedbackResponse> {
    logger.log('FeedbackService: Submitting feedback', request.feedback_type);
    return api.request<FeedbackResponse>('/feedback', {
      method: 'POST',
      body: request,
    });
  }

  /**
   * Submit positive feedback (thumbs up)
   */
  async thumbsUp(
    conversationId: string,
    messageId: string,
    userQuery?: string,
    aiResponse?: string,
    memoriesUsed?: string[]
  ): Promise<FeedbackResponse> {
    return this.submitFeedback({
      feedback_type: 'positive',
      feedback_context: 'response',
      conversation_id: conversationId,
      message_id: messageId,
      user_query: userQuery,
      ai_response: aiResponse,
      memories_used: memoriesUsed,
    });
  }

  /**
   * Submit negative feedback (thumbs down)
   */
  async thumbsDown(
    conversationId: string,
    messageId: string,
    userQuery?: string,
    aiResponse?: string,
    memoriesUsed?: string[]
  ): Promise<FeedbackResponse> {
    return this.submitFeedback({
      feedback_type: 'negative',
      feedback_context: 'response',
      conversation_id: conversationId,
      message_id: messageId,
      user_query: userQuery,
      ai_response: aiResponse,
      memories_used: memoriesUsed,
    });
  }

  /**
   * Submit a correction
   */
  async submitCorrection(
    conversationId: string,
    messageId: string,
    correctionText: string,
    userQuery?: string,
    aiResponse?: string
  ): Promise<FeedbackResponse> {
    return this.submitFeedback({
      feedback_type: 'correction',
      feedback_context: 'response',
      conversation_id: conversationId,
      message_id: messageId,
      user_query: userQuery,
      ai_response: aiResponse,
      correction_text: correctionText,
    });
  }

  /**
   * Get learned user preferences
   */
  async getPreferences(): Promise<UserPreferences> {
    logger.log('FeedbackService: Getting user preferences');
    return api.request<UserPreferences>('/feedback/preferences');
  }

  /**
   * Get learning statistics
   */
  async getStats(): Promise<LearningStats> {
    logger.log('FeedbackService: Getting learning stats');
    return api.request<LearningStats>('/feedback/stats');
  }

  /**
   * Log a memory access (for tracking patterns)
   */
  async logMemoryAccess(
    memoryId: string,
    accessType: 'search' | 'chat_retrieval' | 'connection' | 'direct' | 'suggestion' = 'direct',
    queryText?: string
  ): Promise<void> {
    logger.log('FeedbackService: Logging memory access', memoryId);
    await api.request(`/feedback/memory/${memoryId}/accessed?access_type=${accessType}${queryText ? `&query_text=${encodeURIComponent(queryText)}` : ''}`, {
      method: 'POST',
    });
  }
}

export const feedbackService = new FeedbackService();

File: backend/app/api/feedback.py
"""Feedback API - Endpoints for user feedback and adaptive learning."""

from uuid import UUID
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.api.deps import CurrentUser, Database
from app.services.adaptive_learning_service import AdaptiveLearningService

router = APIRouter(prefix="/feedback", tags=["feedback"])


# ==================== SCHEMAS ====================

class FeedbackRequest(BaseModel):
    """Request schema for submitting feedback."""
    feedback_type: str  # 'positive', 'negative', 'correction'
    feedback_context: str  # 'response', 'suggestion', 'memory_retrieval', 'action'
    conversation_id: Optional[str] = None
    message_id: Optional[str] = None
    user_query: Optional[str] = None
    ai_response: Optional[str] = None
    correction_text: Optional[str] = None
    memories_used: Optional[list[str]] = None


class FeedbackResponse(BaseModel):
    """Response schema for feedback submission."""
    id: str
    feedback_type: str
    message: str


class PreferencesResponse(BaseModel):
    """Response schema for user preferences."""
    preferences: dict
    user_model_prompt: str


class LearningStatsResponse(BaseModel):
    """Response schema for learning statistics."""
    memories: dict
    feedback: dict
    preferences_learned: int
    insights_generated: int


# ==================== ENDPOINTS ====================

@router.post("", response_model=FeedbackResponse)
async def submit_feedback(
    request: FeedbackRequest,
    user: CurrentUser,
    db: Database,
):
    """
    Submit feedback on an AI response.

    This feedback is used to:
    - Reinforce or penalize memory importance
    - Learn user preferences
    - Improve future responses

    Feedback types:
    - positive: The response was helpful
    - negative: The response was not helpful
    - correction: The response needed correction (provide correction_text)
    """
    service = AdaptiveLearningService(db)

    feedback = await service.record_feedback(
        user_id=user.id,
        feedback_type=request.feedback_type,
        feedback_context=request.feedback_context,
        conversation_id=request.conversation_id,
        message_id=request.message_id,
        user_query=request.user_query,
        ai_response=request.ai_response,
        correction_text=request.correction_text,
        memories_used=request.memories_used,
    )

    return FeedbackResponse(
        id=str(feedback.id),
        feedback_type=feedback.feedback_type,
        message="Feedback recorded successfully",
    )


@router.get("/preferences", response_model=PreferencesResponse)
async def get_user_preferences(
    user: CurrentUser,
    db: Database,
):
    """
    Get learned user preferences.

    Returns preferences organized by type:
    - communication_style: How the user prefers to communicate
    - interests: Topics the user is interested in
    - behavior_pattern: Behavioral patterns observed
    - correction_learning: Things learned from corrections
    """
    service = AdaptiveLearningService(db)

    preferences = await service.get_user_preferences(user.id)
    user_model_prompt = await service.get_user_model_prompt(user.id)

    return PreferencesResponse(
        preferences=preferences,
        user_model_prompt=user_model_prompt,
    )


@router.get("/stats", response_model=LearningStatsResponse)
async def get_learning_stats(
    user: CurrentUser,
    db: Database,
):
    """
    Get statistics about the adaptive learning system.

    Returns:
    - Memory stats (count, average strength, emotional weight)
    - Feedback stats (positive/negative counts)
    - Number of preferences learned
    - Number of insights generated
    """
    service = AdaptiveLearningService(db)
    stats = await service.get_learning_stats(user.id)

    return LearningStatsResponse(**stats)


@router.post("/memory/{memory_id}/accessed")
async def log_memory_access(
    memory_id: UUID,
    user: CurrentUser,
    db: Database,
    access_type: str = "direct",
    query_text: Optional[str] = None,
):
    """
    Log a memory access (for tracking access patterns).

    Access types:
    - search: Retrieved via search
    - chat_retrieval: Retrieved during chat
    - connection: Retrieved via connection
    - direct: Directly accessed
    - suggestion: Used in suggestion
    """
    service = AdaptiveLearningService(db)

    await service.log_memory_access(
        user_id=user.id,
        memory_id=memory_id,
        access_type=access_type,
        query_text=query_text,
    )

    return {"message": "Memory access logged"}


@router.post("/insights/extract")
async def extract_insights(
    user: CurrentUser,
    db: Database,
    days: int = 30,
):
    """
    Manually trigger insight extraction from recent memories.

    This analyzes memories from the last N days and extracts patterns.
    Normally runs automatically via background job.
    """
    service = AdaptiveLearningService(db)
    insights = await service.extract_patterns(user.id, days=days)

    return {
        "insights_extracted": len(insights),
        "insights": [
            {
                "id": str(i.id),
                "type": i.insight_type,
                "title": i.title,
                "content": i.content,
                "confidence": i.confidence,
            }
            for i in insights
        ],
    }
